File: src/clickhouse/access.rs
//! Who can do what.
//!
//! Read-only, deliberately. Flint will show you that a user can log in without
//! a password, or that a role nobody holds is still carrying grants — it will
//! not change any of it. Granting and revoking are decisions with consequences
//! that outlive a click, and they belong in a statement somebody wrote on
//! purpose.
//!
//! Every one of these tables is privileged: a read-only role usually cannot
//! read `system.users` at all. That is a configuration fact and is reported as
//! one, with the grant it would need.

use serde::{Deserialize, Serialize};

use super::{Client, Reach};
use crate::error::Result;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub name: String,
    /// An array: recent ClickHouse lets one user hold several authentication
    /// methods, and `no_password` among them is the fact worth seeing.
    #[serde(default)]
    pub auth_type: Vec<String>,
    #[serde(default)]
    pub host_ip: Vec<String>,
    #[serde(default)]
    pub host_names: Vec<String>,
    pub default_roles_all: bool,
    #[serde(default)]
    pub default_roles_list: Vec<String>,
    #[serde(default)]
    pub default_database: String,
    /// One entry per authentication method — a user may hold several, each
    /// with its own expiry. Epoch means "no expiry", which is how ClickHouse
    /// says it.
    #[serde(default)]
    pub valid_until: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Role {
    pub name: String,
    #[serde(default)]
    pub storage: String,
}

/// One scope somebody has rights on: `analytics.*`, `system.query_log`, `*.*`.
///
/// Aggregated by scope rather than listed per access type, because ClickHouse
/// stores one row per privilege and a full-access user has seventy of them.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Grant {
    pub grantee: String,
    pub is_user: bool,
    /// `*` where the grant is not narrowed to one.
    pub database: String,
    pub table: String,
    pub with_grant_option: bool,
    /// A partial revoke: rights taken back inside a wider grant. Rare, and
    /// invisible if you only read the positive rows.
    pub revoked: bool,
    pub access: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleGrant {
    pub grantee: String,
    pub is_user: bool,
    pub role: String,
    pub is_default: bool,
    pub with_admin_option: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct AccessReport {
    pub available: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    pub users: Vec<User>,
    pub roles: Vec<Role>,
    pub grants: Vec<Grant>,
    pub role_grants: Vec<RoleGrant>,
}

fn unavailable(reason: String) -> AccessReport {
    AccessReport {
        available: false,
        reason: Some(reason),
        users: Vec::new(),
        roles: Vec::new(),
        grants: Vec::new(),
        role_grants: Vec::new(),
    }
}

pub async fn access(ch: &Client) -> Result<AccessReport> {
    match ch.reach("users").await? {
        Reach::Readable => {}
        Reach::Denied => {
            return Ok(unavailable(
                "this user cannot read system.users. Reading access control needs SHOW USERS, \
                 SHOW ROLES and SELECT on those tables — none of which lets Flint change \
                 anything."
                    .into(),
            ))
        }
        Reach::Absent => return Ok(unavailable("this server has no system.users".into())),
    }

    let valid_until = ch.col_or("users", "valid_until", "[]").await?;
    let default_database = ch.col_or("users", "default_database", "''").await?;

    let users: Vec<User> = ch
        .rows(&format!(
            "SELECT name                                 AS name, \
                    auth_type                            AS auth_type, \
                    host_ip                              AS host_ip, \
                    host_names                           AS host_names, \
                    CAST(default_roles_all != 0 AS Bool) AS default_roles_all, \
                    default_roles_list                   AS default_roles_list, \
                    {default_database}                   AS default_database, \
                    arrayMap(x -> toString(x), {valid_until}) AS valid_until \
             FROM system.users \
             ORDER BY name \
             LIMIT 2000"
        ))
        .await?;

    // Each of the remaining three can be denied on its own — a role granted
    // SHOW USERS is not necessarily granted SHOW ROLES — so a failure on any of
    // them costs that list, not the page.
    let roles: Vec<Role> = optional(
        ch,
        "SELECT name AS name, storage AS storage FROM system.roles ORDER BY name LIMIT 2000",
    )
    .await?;

    let grants: Vec<Grant> = optional(
        ch,
        "SELECT coalesce(user_name, role_name)       AS grantee, \
                CAST(user_name IS NOT NULL AS Bool)  AS is_user, \
                coalesce(database, '*')              AS database, \
                coalesce(table, '*')                 AS table, \
                CAST(grant_option != 0 AS Bool)      AS with_grant_option, \
                CAST(is_partial_revoke != 0 AS Bool) AS revoked, \
                arraySort(groupArray(access_type))   AS access \
         FROM system.grants \
         GROUP BY grantee, is_user, database, table, with_grant_option, revoked \
         ORDER BY grantee, database, table \
         LIMIT 5000",
    )
    .await?;

    let role_grants: Vec<RoleGrant> = optional(
        ch,
        "SELECT coalesce(user_name, role_name)             AS grantee, \
                CAST(user_name IS NOT NULL AS Bool)        AS is_user, \
                granted_role_name                          AS role, \
                CAST(granted_role_is_default != 0 AS Bool) AS is_default, \
                CAST(with_admin_option != 0 AS Bool)       AS with_admin_option \
         FROM system.role_grants \
         ORDER BY grantee, role \
         LIMIT 5000",
    )
    .await?;

    Ok(AccessReport {
        available: true,
        reason: None,
        users,
        roles,
        grants,
        role_grants,
    })
}

/// A denied list is empty, not fatal.
async fn optional<T: serde::de::DeserializeOwned>(ch: &Client, sql: &str) -> Result<Vec<T>> {
    match ch.rows::<T>(sql).await {
        Ok(rows) => Ok(rows),
        Err(
            e @ (crate::error::Error::ClickHouse { code: 497, .. }
            | crate::error::Error::ClickHouse { code: 164, .. }),
        ) => {
            tracing::debug!("access list unavailable: {e}");
            Ok(Vec::new())
        }
        Err(e) => Err(e),
    }
}

File: frontend/src/lib/access.ts
/** Who can do what.
 *
 *  A dump of `system.grants` is unreadable — ClickHouse stores one row per
 *  privilege, so a full-access user is seventy rows. What a reader wants is the
 *  scopes, and then the handful of facts worth knowing about them: who can log
 *  in without a password, who can grant to others, which role nobody holds.
 *
 *  Read-only by design. Flint will point these out; changing them is a decision
 *  that belongs in a statement somebody wrote on purpose. */

export interface User {
  name: string
  /** Where the account is defined. `local_directory` is what SQL writes;
   *  `users_xml` is a file, and SQL cannot touch it. */
  storage: string
  auth_type: string[]
  host_ip: string[]
  host_names: string[]
  default_roles_all: boolean
  default_roles_list: string[]
  default_database: string
  /** One per authentication method; epoch means no expiry. */
  valid_until: string[]
}

export interface Role {
  name: string
  storage: string
}

export interface Grant {
  grantee: string
  is_user: boolean
  database: string
  table: string
  with_grant_option: boolean
  revoked: boolean
  access: string[]
}

export interface RoleGrant {
  grantee: string
  is_user: boolean
  role: string
  is_default: boolean
  with_admin_option: boolean
}

export interface AccessReport {
  available: boolean
  reason?: string
  users: User[]
  roles: Role[]
  grants: Grant[]
  role_grants: RoleGrant[]
  /** Every privilege this server understands, for the grant form to offer.
   *  Empty where `system.privileges` could not be read, in which case the form
   *  takes free text and lets ClickHouse do the refusing. */
  privileges: string[]
}

/** `analytics.*`, `system.query_log`, `everything`. */
export function scopeOf(grant: Grant): string {
  if (grant.database === '*') return 'everything'
  if (grant.table === '*') return `${grant.database}.*`
  return `${grant.database}.${grant.table}`
}

/** The privileges, shortened where there are too many to read.
 *
 *  A grantee with everything has seventy access types; listing them all says
 *  less than saying there are seventy. */
export function accessOf(grant: Grant, show = 4): string {
  if (grant.access.length <= show) return grant.access.join(', ')
  return `${grant.access.slice(0, show).join(', ')} and ${grant.access.length - show} more`
}

export interface Note {
  /** `watch` is a fact worth seeing; `alarm` is one somebody should act on. */
  level: 'alarm' | 'watch'
  says: string
}

/** What is worth pointing out about one user.
 *
 *  Deliberately few. Every self-hosted ClickHouse has every user reachable from
 *  anywhere, so saying that about each of them is noise — it is only worth
 *  saying beside something else. */
export function notesForUser(
  user: User,
  grants: Grant[],
  roleGrants: RoleGrant[],
): Note[] {
  const notes: Note[] = []
  const mine = grants.filter((g) => g.is_user && g.grantee === user.name)
  const roles = roleGrants.filter((r) => r.is_user && r.grantee === user.name)

  if (user.auth_type.includes('no_password')) {
    notes.push({
      level: 'alarm',
      says: 'can connect with no password at all',
    })
  }
  if (mine.some((g) => g.database === '*' && !g.revoked)) {
    notes.push({ level: 'watch', says: 'has rights on everything' })
  }
  if (mine.some((g) => g.with_grant_option)) {
    notes.push({ level: 'watch', says: 'can grant its rights to others' })
  }
  if (roles.some((r) => r.with_admin_option)) {
    notes.push({ level: 'watch', says: 'can grant its roles to others' })
  }
  if (!mine.length && !roles.length) {
    notes.push({
      level: 'watch',
      says: 'has no rights and no roles — it can connect and see nothing',
    })
  }
  // Epoch is ClickHouse's way of saying "never expires", and there is one
  // entry per authentication method — the earliest real one is the one that
  // matters.
  const expiries = user.valid_until.filter((v) => v && !v.startsWith('1970')).sort()
  if (expiries.length) {
    notes.push({ level: 'watch', says: `expires ${expiries[0]!.slice(0, 16)}` })
  }
  return notes
}

/** A role nobody holds is carrying grants for nobody: dead configuration that
 *  still looks like policy. */
export function notesForRole(role: Role, grants: Grant[], roleGrants: RoleGrant[]): Note[] {
  const notes: Note[] = []
  const held = roleGrants.some((r) => r.role === role.name)
  const mine = grants.filter((g) => !g.is_user && g.grantee === role.name)
  if (!held) {
    notes.push({ level: 'watch', says: 'nobody holds this role' })
  }
  if (!mine.length) {
    notes.push({ level: 'watch', says: 'it grants nothing' })
  }
  if (mine.some((g) => g.database === '*' && !g.revoked)) {
    notes.push({ level: 'watch', says: 'it grants rights on everything' })
  }
  return notes
}

export function grantsFor(report: AccessReport, name: string, isUser: boolean): Grant[] {
  return report.grants.filter((g) => g.is_user === isUser && g.grantee === name)
}

export function rolesFor(report: AccessReport, name: string): RoleGrant[] {
  return report.role_grants.filter((r) => r.is_user && r.grantee === name)
}

export function holdersOf(report: AccessReport, role: string): string[] {
  return report.role_grants.filter((r) => r.role === role).map((r) => r.grantee)
}
